brainiak/utils/fmrisim.py

```diff
@@ -519,8 +664,247 @@ def apply_signal(signal_function,
+    The snr of the volume (how much greater in activity the brain is from
+    the background)
+
+    """
+
+    # What are the midpoints to be extracted
+    mid_x_idx = int(np.ceil(volume.shape[0] / 2))
+    mid_tr_idx = int(np.ceil(volume.shape[3] / 2))
+
+    # Pull out the slices
+    slice_volume = volume[mid_x_idx, :, :, mid_tr_idx]
+    slice_mask = mask[mid_x_idx, :, :, mid_tr_idx]
+
+    # Divide by the mask (if the middle slice was low in grey matter mass
+    # then you would have a lower mean signal by default)
+    mean_signal = (slice_volume[slice_mask > 0] / slice_mask[slice_mask >
+                                                             0]).mean()
+
+    # What is the overall noise
+    overall_noise = float(slice_volume[slice_mask == 0].std())
+
+    # Calculate snr
+    snr = mean_signal / overall_noise
+
+    # Convert from memmap
+    snr = float(snr)
+
+    return overall_noise, snr
+
+
+def _calc_autoregression(volume,
+                         mask,
+                         ar_order=2,
+                         ):
+    """ Calculate the autoregressive sigma of the data.
+
+    Parameters
+    ----------
+    volume : 4d masarray, float
+    Input data to be calculate the autoregression
+
+    mask : 4d array, float
+    What voxels of the input are within the brain
+
+    ar_order : int
+    What order of the autoregression do you want to pull out
+
+    Returns
+    -------
+    float
+    A sigma of the autoregression in the data
+
+    """
+
+    # Calculate the time course
+    timecourse = np.mean(volume[mask[:, :, :, 0] > 0], 0)
+
+    # Pull out the AR values (depends on order)
+    auto_reg_sigma = ar.AR_est_YW(timecourse, ar_order)[0][1]
+
+    # What is the size of the change in the time course
+    drift_sigma = timecourse.std().tolist()
+
+    return auto_reg_sigma, drift_sigma
+
+
+def calc_noise(volume,
+               mask=None,
+               ):
+    """ Calculates the noise properties of the volume supplied.
+    This estimates what noise properties the volume has. For instance it
+    determines the spatial smoothness, the autoregressive noise, system
+    noise etc. Read the doc string for generate_noise to understand how
+    these different types of noise interact.
+
+    Parameters
+    ----------
+    volume : str or 4d numpy array, float
+        Take in a functional volume (either the file name or the numpy
+        array) to be used to estimate the noise properties of this
+
+    mask : 4d numpy array, float
+        The mask to be used, the same size as the volume
+
+    Returns
+    -------
+
+    dict
+    Return a dictionary of the calculated noise parameters of the provided
+    dataset
+
+    """
+
+    # Preset
+    noise_dict = {}
+
+    # If a string name is provided then load this file and pull out the
+    # nifti into a numpy array
+    if isinstance(volume, str):
+        nifti = nibabel.load(volume)  # Load the file
+        volume = nifti.get_data()  # Pull out the data
```

Comment:
It is indeed. However, so far we have only been using it in examples, not in BrainIAK code. I would prefer to not add it as a BrainIAK dependency yet, if it is not really necessary. The way I see it, here it is not necessary, because you can load the file outside BrainIAK, like you do now in the example.